R/position-jitterbin.r

```diff
@@ -0,0 +1,79 @@
+#' Adjust position by jittering inside the data range.
+#'
+#' This produces nicer plots in some situations, especially for
+#' logistic regression plots. The default values for
+#' \code{width} and \code{height} have been tailored for
+#' plots of binary data against a predictor. The former defaults to
+#' zero, while the latter default to 5\% of the data resolution (0.05
+#' in case of a 0-1 binary variable).
+#'
+#' @family position adjustments
+#' @param width degree of jitter in x direction. Defaults to 5\% of
+#' the resolution of the data.
+#' @param height degree of jitter in y direction. Defaults to 0.
+#' @param bounded_scales scales whose ranges bound the jitter. Can be
+#' \code{"x"}, \code{"y"} or \code{c("x", "y")}. Defaults to
+#' \code{"y"}.
+#' @export
+#' @examples
+#' movies_sub <- movies[sample(nrow(movies), 1000), ]
+#' ggplot(movies_sub, aes(year, Comedy)) +
+#'   geom_point(position = position_jitterbin(), alpha = 0.5) +
+#'   stat_smooth(method = "glm", family = "binomial", level = 0)
+position_jitterbin <- function (width = NULL, height = NULL, bounded_scales = "y") {
+  PositionJitterbin$new(width = width, height = height, bounded_scales = bounded_scales)
+}
+
+PositionJitterbin <- proto(Position, {
+  width <- NULL
+  height <- NULL
+  bounded_scales <- NULL
+
+  new <- function(., width = NULL, height = NULL, bounded_scales = "y") {
+    .$proto(width = width, height = height, bounded_scales = bounded_scales)
+  }
+
+  objname <- "position_jitterbin"
+
+  adjust <- function(., data) {
+    if (empty(data)) return(data.frame())
+    check_required_aesthetics(c("x", "y"), names(data), "position_jitterbin")
+
+    if (is.null(.$width)) .$width <- 0
+    if (is.null(.$height)) .$height <- resolution(data$y, zero = FALSE) * 0.05
+
+    bound_x <- "x" %in% .$bounded_scales
+    bound_y <- "y" %in% .$bounded_scales
+    if ((bound_x && !length(unique(data$x)) == 2) ||
+        (bound_y && !length(unique(data$y)) == 2))
+      stop("position_jitterbin can only bound binary scales")
+
+    trans_x <- NULL
+    trans_y <- NULL
+    if(.$width > 0) {
+      trans_x <- function(x) {
```

Comment:
I think this could probably be much simplified by using `runif()` in both branches, and just adjusting the min and max values.

Alternatively, another parameterisation would be to alway jitter towards the centre (but that might pose just as many problems in other cases)